### ISSUE TITLE
Maybe bounds for recursive types can be improved

### ISSUE THREAD

**Issue by thiolliere** [paritytech/parity-scale-codec#217]
This is a potential improvement but it doesn't completely solve the issue for recursive types so not sure it matters, I'm OK to close.

the following example doesn't compiles:
```rust
#[test]
fn recursive_variant_3_encode_works() {
	#[derive(Default)]
	struct NonEncodableType;

	#[derive(Default, Encode)]
	struct Recursive<A, B, N> {
		other: Vec<(A, B, Recursive<A, B, N>)>,
		n: core::marker::PhantomData<N>,
	}

	let val: Recursive<u32, i32, NonEncodableType> = Recursive::default();
	val.encode();
}
```
this is because the bounds are:
```
A: Encode,
B: Encode,
N: Encode,
core::marker::PhantomData<N>: Encode,
```

for recursive types like `Vec<(A, B, Recursive<A, B, N>)>,` I think we could only bound all used generics but not those inside Self types: thus `A` and `B` here.

(I tested and bounding only A, B, and phantom make the test works, though I haven't tried substrate codebase)

Though anyway this is just best effort, ideally we should bound `Vec<(A, B, Recursive<A, B, N>)>: Encode` but this is not possible.

**Comment by bkchr** [comment]
> Though anyway this is just best effort, ideally we should bound `Vec<(A, B, Recursive<A, B, N>)>: Encode` but this is not possible.

Are you sure we not already doing this?

**Comment by thiolliere** [comment]
> > Though anyway this is just best effort, ideally we should bound `Vec<(A, B, Recursive<A, B, N>)>: Encode` but this is not possible.
> 
> Are you sure we not already doing this?

Yes, becauce if we bound some types which contains Self then compiler will error with: `overflow evaluating the requirement..`

So currently for type:
```rust
	struct Recursive<N> {
		other: Vec<Recursive<N>>,
		phantom: core::marker::PhantomData<N>,
	}
```
we bound `N` and `PhantomData<N>` but I suggest only bounding `PhantomData<N>`

though bound for recursive type are very tricky thus I'm not sure it worth trying to solve what nobody complain about.

**Comment by bkchr** [comment]
Ahh yeah, I understand. Does rust accepts it without a bound on `N`?
